File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/individuals/designation_controller.py
import logging
import inspect
from django.contrib import messages
from django.contrib.auth.decorators import login_required
from django.core.exceptions import PermissionDenied
from django.http import HttpRequest, HttpResponse
from django.shortcuts import get_object_or_404, redirect, render
from django.utils.decorators import method_decorator
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller
from ....Models.companies.department import Department
from ....Models.individuals.designation import Designation

logger = logging.getLogger(__name__)

class DesignationController(Controller):

    @method_decorator(login_required)
    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('manage designation'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: manage designation'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            designations = Designation.objects.filter(created_by=request.user.creator_id())
            return render(request, 'designation/index.html', {'designations': designations})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('create designation'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: create designation'),
                    ref=f'{CN}::{MN}',
                    logger=logger,
                    json={}
                )
            departments_qs = Department.objects.filter(created_by=request.user.creator_id())
            departments    = {dept.id: dept.name for dept in departments_qs}
            return render(request, 'designation/create.html', {'departments': departments})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger, json={})

    @method_decorator(login_required)
    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('create designation'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: create designation'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            dept_id = request.POST.get('department_id')
            name    = request.POST.get('name')
            if not dept_id or not name:
                messages.error(request, 'Department and name are required.')
                return redirect(get_redirect_url(request))
            if len(name) > 20:
                messages.error(request, 'Name must be at most 20 characters.')
                return redirect(get_redirect_url(request))
            designation = Designation(
                department_id=dept_id,
                name=name,
                created_by=request.user.creator_id()
            )
            designation.save()
            messages.success(request, 'Designation successfully created.')
            return redirect('designation.index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def show(cls, request: HttpRequest, designation_id: int) -> HttpResponse:
        # CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        # TODO No detail view; redirect back to list
        return redirect('designation.index')

    @method_decorator(login_required)
    @classmethod
    def edit(cls, request: HttpRequest, designation_id: int) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            designation = get_object_or_404(Designation, pk=designation_id)
            if not request.user.has_perm('edit designation'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: edit designation'),
                    ref=f'{CN}::{MN}',
                    logger=logger,
                    json={}
                )
            if designation.created_by != request.user.creator_id():
                return default_permission_denial(
                    request,
                    err=PermissionDenied('Permission denied.'),
                    ref=f'{CN}::{MN}',
                    logger=logger,
                    json={}
                )
            dept = Department.objects.filter(id=designation.department_id).first()
            departments = {dept.id: dept.name} if dept else {}
            return render(request, 'designation/edit.html', {
                'designation': designation,
                'departments': departments
            })
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger, json={})

    @method_decorator(login_required)
    @classmethod
    def update(cls, request: HttpRequest, designation_id: int) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('edit designation'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: edit designation'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            designation = get_object_or_404(Designation, pk=designation_id)
            if designation.created_by != request.user.creator_id():
                return default_permission_denial(
                    request,
                    err=PermissionDenied('Permission denied.'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            dept_id = request.POST.get('department_id')
            name    = request.POST.get('name')
            if not dept_id or not name:
                messages.error(request, 'Department and name are required.')
                return redirect(get_redirect_url(request))
            if len(name) > 20:
                messages.error(request, 'Name must be at most 20 characters.')
                return redirect(get_redirect_url(request))
            designation.department_id = dept_id
            designation.name          = name
            designation.save()
            messages.success(request, 'Designation successfully updated.')
            return redirect('designation.index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def destroy(cls, request: HttpRequest, designation_id: int) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('delete designation'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: delete designation'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            designation = get_object_or_404(Designation, pk=designation_id)
            if designation.created_by != request.user.creator_id():
                return default_permission_denial(
                    request,
                    err=PermissionDenied('Permission denied.'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            designation.delete()
            messages.success(request, 'Designation successfully deleted.')
            return redirect('designation.index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

File: _inc/laravel/app/Http/Controllers/Individuals/DesignationController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    CompaniesConstants,
    DatabaseConstants,
    UsersConstants,
    ViewsConstants
};
use App\Models\{Department, Designation};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{RedirectResponse, Request};
use Illuminate\Support\Facades\{Log, Validator, View as ViewFacade};
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\HasCrudConstants;
class DesignationController extends Controller
{
    use HasCrudConstants;

    use ChecksLogin, ChecksPermissions;

    public function index(Request $request): View|RedirectResponse
    {
        $action = 'DesignationController@index';
        $view   = ViewsConstants::DSG . '.index';

        return $this->measureProfile($action, function () use ($request, $action, $view) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;

            // authorize
            $t = microtime(true);
            try {
                $this->_authorize($request, 'manage designation');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');
            } catch (AuthorizationException $e) {
                $this->logExecutionTime($t, $action . '::authorize', 'denied');
                return defaultPermissionDenial($request, $e, $action);
            }

            // query
            $t = microtime(true);
            $designations = Designation::where(DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())->get();
            $this->logExecutionTime($t, $action . '::query', 'rows: ' . $designations->count());

            // view check
            $t = microtime(true);
            if (!ViewFacade::exists($view)) {
                $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                return defaultUndefinedException($request, new \RuntimeException("View not found: $view"), $action);
            }
            $this->logExecutionTime($t, $action . '::viewCheck', 'exists');

            return view($view, compact('designations'));
        }, ['uri' => $request->getRequestUri()]);
    }

    public function create(Request $request): View|RedirectResponse
    {
        $action = 'DesignationController@create';
        $view   = ViewsConstants::DSG . '.create';

        return $this->measureProfile($action, function () use ($request, $action, $view) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            // authorize
            $t = microtime(true);
            try {
                $this->_authorize($request, 'create designation');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');
            } catch (AuthorizationException $e) {
                $this->logExecutionTime($t, $action . '::authorize', 'denied');
                return defaultPermissionDenial($request, $e, $action);
            }

            // load form data
            $t = microtime(true);
            $departmentList = Department::where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()->creatorId())
                ->pluck(CompaniesConstants::COL_DEP_NM, 'id');
            $this->logExecutionTime($t, $action . '::loadFormData', 'deps: ' . $departmentList->count());

            // view check
            $t = microtime(true);
            if (!ViewFacade::exists($view)) {
                $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                return defaultUndefinedException($request, new \RuntimeException("View not found: $view"), $action);
            }
            $this->logExecutionTime($t, $action . '::viewCheck', 'exists');

            return view($view, compact('departmentList'));
        }, ['uri' => $request->getRequestUri()]);
    }

    public function store(Request $request): RedirectResponse
    {
        $action = 'DesignationController@store';

        return $this->measureProfile($action, function () use ($request, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            // authorize
            $t = microtime(true);
            try {
                $this->_authorize($request, 'create designation');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');
            } catch (AuthorizationException $e) {
                $this->logExecutionTime($t, $action . '::authorize', 'denied');
                return defaultPermissionDenial($request, $e, $action);
            }

            // validate (use 'name' from request body; dep id from constant)
            $t = microtime(true);
            $validator = Validator::make($request->all(), [
                CompaniesConstants::COL_DEP_ID => 'required',
                'name'                         => 'required|max:20',
            ]);
            if ($validator->fails()) {
                $this->logExecutionTime($t, $action . '::validate', 'failed');
                return redirect()->back()->with('error', $validator->getMessageBag()->first());
            }
            $this->logExecutionTime($t, $action . '::validate', 'ok');

            // persist
            $t = microtime(true);
            $user = $request->user();
            $designation = Designation::create([
                CompaniesConstants::COL_DEP_ID   => $request->input(CompaniesConstants::COL_DEP_ID),
                'name'                           => $request->input('name'),
                DatabaseConstants::COL_TABLE_CREATOR => $user?->creatorId()
            ]);
            $this->logExecutionTime($t, $action . '::persist', 'id: ' . $designation->id);
            Log::info('Designation created', ['id' => $designation->id]);

            return redirect()->route(ViewsConstants::DSG . '.index')
                ->with('success', __('Designation successfully created.'));
        }, ['uri' => $request->getRequestUri()]);
    }

    public function show(): RedirectResponse
    {
        // simple redirect; profiling not necessary
        return redirect()->route(ViewsConstants::DSG . '.index');
    }

    public function edit(Request $request, Designation $designation): View|RedirectResponse
    {
        $action = 'DesignationController@edit';
        $view   = ViewsConstants::DSG . '.edit';

        return $this->measureProfile($action, function () use ($request, $designation, $action, $view) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            // authorize
            $t = microtime(true);
            try {
                $this->_authorize($request, 'edit designation');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');
            } catch (AuthorizationException $e) {
                $this->logExecutionTime($t, $action . '::authorize', 'denied');
                return defaultPermissionDenial($request, $e, $action);
            }

            // ownership check
            $t = microtime(true);
            if ($designation->created_by !== $request->user()->creatorId()) {
                $this->logExecutionTime($t, $action . '::owner', 'denied');
                return defaultPermissionDenial($request, new AuthorizationException(), $action);
            }
            $this->logExecutionTime($t, $action . '::owner', 'ok');

            // load data
            $t = microtime(true);
            $departmentList = Department::where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()->creatorId())
                ->pluck(CompaniesConstants::COL_DEP_NM, 'id');
            $this->logExecutionTime($t, $action . '::loadFormData', 'deps: ' . $departmentList->count());

            // view check
            $t = microtime(true);
            if (!ViewFacade::exists($view)) {
                $this->logExecutionTime($t, $action . '::viewCheck', 'missing');
                return defaultUndefinedException($request, new \RuntimeException("View not found: $view"), $action);
            }
            $this->logExecutionTime($t, $action . '::viewCheck', 'exists');

            return view($view, compact('designation', 'departmentList'));
        }, ['uri' => $request->getRequestUri(), 'id' => $designation->id]);
    }

    public function update(Request $request, Designation $designation): RedirectResponse
    {
        $action = 'DesignationController@update';

        return $this->measureProfile($action, function () use ($request, $designation, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            // authorize
            $t = microtime(true);
            try {
                $this->_authorize($request, 'edit designation');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');
            } catch (AuthorizationException $e) {
                $this->logExecutionTime($t, $action . '::authorize', 'denied');
                return defaultPermissionDenial($request, $e, $action);
            }

            // ownership check
            $t = microtime(true);
            if ($designation->created_by !== $request->user()->creatorId()) {
                $this->logExecutionTime($t, $action . '::owner', 'denied');
                return defaultPermissionDenial($request, new AuthorizationException(), $action);
            }
            $this->logExecutionTime($t, $action . '::owner', 'ok');

            // validate
            $t = microtime(true);
            $validator = Validator::make($request->all(), [
                CompaniesConstants::COL_DEP_ID => 'required',
                'name'                         => 'required|max:20'
            ]);
            if ($validator->fails()) {
                $this->logExecutionTime($t, $action . '::validate', 'failed');
                return redirect()->back()->with('error', $validator->getMessageBag()->first());
            }
            $this->logExecutionTime($t, $action . '::validate', 'ok');

            // persist
            $t = microtime(true);
            $designation->update([
                CompaniesConstants::COL_DEP_ID => $request->input(CompaniesConstants::COL_DEP_ID),
                'name'                         => $request->input('name')
            ]);
            $this->logExecutionTime($t, $action . '::persist', 'id: ' . $designation->id);

            return redirect()->route(ViewsConstants::DSG . '.index')
                ->with('success', __('Designation successfully updated.'));
        }, ['uri' => $request->getRequestUri(), 'id' => $designation->id]);
    }

    public function destroy(Request $request, Designation $designation): RedirectResponse
    {
        $action = 'DesignationController@destroy';

        return $this->measureProfile($action, function () use ($request, $designation, $action) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;

            // authorize
            $t = microtime(true);
            try {
                $this->_authorize($request, 'delete designation');
                $this->logExecutionTime($t, $action . '::authorize', 'ok');
            } catch (AuthorizationException $e) {
                $this->logExecutionTime($t, $action . '::authorize', 'denied');
                return defaultPermissionDenial($request, $e, $action);
            }

            // ownership check
            $t = microtime(true);
            if ($designation->created_by !== $request->user()->creatorId()) {
                $this->logExecutionTime($t, $action . '::owner', 'denied');
                return defaultPermissionDenial($request, new AuthorizationException(), $action);
            }
            $this->logExecutionTime($t, $action . '::owner', 'ok');

            // delete
            $t = microtime(true);
            $designation->delete();
            $this->logExecutionTime($t, $action . '::delete', 'id: ' . $designation->id);

            return redirect()->route(ViewsConstants::DSG . '.index')
                ->with('success', __('Designation successfully deleted.'));
        }, ['uri' => $request->getRequestUri(), 'id' => $designation->id]);
    }

    /**
     * Authorize the current user for a given ability.
     *
     * @throws AuthorizationException
     */
    private function _authorize(Request $request, string $ability): void
    {
        $result = self::guard($request, $ability);
        if ($result !== true) {
            throw new AuthorizationException("Unauthorized: {$ability}");
        }
    }
}
